src/node/internal/internal_tls_jsstream.ts

```diff
@@ -0,0 +1,308 @@
+const kPendingShutdownRequest = Symbol('kPendingShutdownRequest');
+const kPendingClose = Symbol('kPendingClose');
+
+/* This class serves as a wrapper for when the C++ side of Node wants access
+ * to a standard JS stream. For example, TLS or HTTP do not operate on network
+ * resources conceptually, although that is the common case and what we are
+ * optimizing for; in theory, they are completely composable and can work with
+ * any stream resource they see.
+ *
+ * For the common case, i.e. a TLS socket wrapping around a net.Socket, we
+ * can skip going through the JS layer and let TLS access the raw C++ handle
+ * of a net.Socket. The flipside of this is that, to maintain composability,
+ * we need a way to create "fake" net.Socket instances that call back into a
+ * "real" JavaScript stream. JSStreamSocket is exactly this.
+ */
+export class JSStreamSocket extends Socket {
+  public stream: Duplex;
+  public [kCurrentWriteRequest]: null | unknown;
+  public [kCurrentShutdownRequest]: null | unknown;
+  public [kPendingShutdownRequest]: null | unknown;
+  public [kPendingClose]: boolean;
+
+  #openPromise: ReturnType<typeof Promise.withResolvers<SocketInfo>>;
+
+  constructor(stream: Duplex) {
+    const closePromise = Promise.withResolvers<void>();
+    const openPromise = Promise.withResolvers<SocketInfo>();
+
+    const webStream = Duplex.toWeb(stream);
+    const writable: Writer = Object.assign({}, webStream.writable, {
+      write: async (data: string | ArrayBufferView): Promise<void> => {
+        stream.write(data);
+      },
+      closed: closePromise.promise,
+      releaseLock: async (): Promise<void> => {},
+    });
+    const handle: Socket['_handle'] = {
+      reading: true,
+      bytesRead: 0,
+      bytesWritten: 0,
+      socket: {
+        startTls(): CloudflareSocket {
+          throw new Error('This should not have triggered');
+        },
+        upgraded: false,
+        secureTransport: 'off',
+        closed: closePromise.promise,
+        close: async (): Promise<void> => {
+          queueMicrotask(() => {
+            closePromise.resolve();
+          });
+          return closePromise.promise;
+        },
+        opened: openPromise.promise,
+        readable: webStream.readable,
+        writable,
+      },
+      reader: new ReadableStreamBYOBReader(webStream.readable),
+      writer: new WritableStreamDefaultWriter<unknown>(writable),
+      options: {
+        host: '0.0.0.0',
+        port: 0,
+        addressType: 4,
+      },
+    };
+
+    stream.pause();
+    stream.on('error', (err) => this.emit('error', err));
+    const ondata = (chunk: string | Buffer): void => {
+      if (typeof chunk === 'string' || stream.readableObjectMode === true) {
+        // Make sure that no further `data` events will happen.
+        stream.pause();
+        stream.removeListener('data', ondata);
+
+        this.emit('error', new ERR_STREAM_WRAP());
```

Comment:
Generally in a Node stream you want to call destroy(error) rather than emitting error yourself

Reply:
This is done to be in par with nodejs behavior